### PR TITLE
Use code_hash for deployed code recognition

### PR DESCRIPTION
**Pull request by mohamedalichelbi**
- Use method on SputnikDAO class to check if deployed contract is SputnikDAO instance
- Integrate the check in a DAO `init()` method. It also fetches & stores DAO `policy` and `last_proposal_id`

### CODE REVIEW

File: src/components/dao/dao.js [hunk 1]
```diff
@@ -146,22 +145,19 @@ export default class Dao extends Component {
...
             noContract,
             noDao,
             noAddProposalRights,
             noApproveProposalRights
         } = this.errors;
 
-        // happens if wallet not logged in
-        if (infos.policy === undefined)
+        // happens if wallet not logged in or DAO object not initialized yet
+        if (dao?.ready === false)
```
Comment (lennczar) [inline src/components/dao/dao.js:148]: if `dao` is not defined / loaded `dao?.ready` will be `null`, not `false`

File: src/components/dao/dao.js [hunk 2]
```diff
@@ -321,99 +317,86 @@ export default class Dao extends Component {
...
-            view(multicall, "get_admins", {}).catch(e => {
-                    if (e.type === "AccountDoesNotExist" && e.toString().includes(` ${multicall} `)) {
-                        noContract.isBad = true;
+        // initialize DAO object
+        SputnikDAO.init(dao_address).catch(e => {})
+        .then((newDAO) => {
+            // DAO not ready => either no SputnikDAO contract on the chosen address
+            // or some error happened during DAO object init.
+            this.setState({ dao: newDAO });
+            if (!newDAO.ready) { return; }
```
Comment (lennczar) [inline src/components/dao/dao.js:333]: Please print something to the console, so we're not left blind if we happen to break this.

Btw, u can merge this `.then` with the one below, no reason to have them seperate.

File: src/components/dao/dao.js [hunk 3]
```diff
@@ -302,10 +302,11 @@ export default class DaoComponent extends Component {
 
     loadInfos() {
 
-        const { addrError, noContract, noDao } = this.errors;
-        const { addr, dao } = this.state;
+        const { addr: addrError, noContract, noDao } = this.errors;
+        const { addr } = this.state;
 
         const multicall = `${addr.value}.${window.nearConfig.MULTICALL_FACTORY_ADDRESS}`;
+        const daoAddress = `${addr.value}.${SputnikDAO.FACTORY_ADDRESS}`;
```
Comment (lennczar) [inline src/components/dao/dao.js:309]: We have `dao` in the state, why not use that?

File: src/components/dao/dao.js [hunk 4]
```diff
@@ -316,7 +317,6 @@ export default class DaoComponent extends Component {
         if (addrError.isBad) {
             noContract.isBad = true;
             noDao.isBad = true;
-            this.setState({ proposed: -1, proposedInfo: {} });
```
Comment (lennczar) [inline src/components/dao/dao.js]: Why remove this line? Its important for two reasons:
- if there is no dao, then there is not creation proposal
- `setState` updates the react component

File: src/components/dao/dao.js [hunk 5]
```diff
@@ -325,44 +325,52 @@ export default class DaoComponent extends Component {
...
-                        if (e.type === "AccountDoesNotExist" && e.toString().includes(` ${multicall} `)) {
-                            noContract.isBad = true;
+            if (!newDAO.ready) {
+                noContract.isBad = true;
+                noDao.isBad = true;
+                this.setState({ loading: false });
+                return;
+            }
+            // DAO correctly initialized, try to fetch multicall info
+            else {
```
Comment (lennczar) [inline src/components/dao/dao.js:340]: No need for `else` if the "then" part already `return`s

File: src/components/dao/dao.js [hunk 6]
```diff
@@ -325,44 +325,52 @@ export default class DaoComponent extends Component {
...
+                Promise.all([
+                    view(multicall, "get_admins", {}).catch(e => {
+                            if (e.type === "AccountDoesNotExist" && e.toString().includes(` ${multicall} `)) {
+                                noContract.isBad = true;
+                            }
                         }
+                    ),
```
Comment (lennczar) [inline src/components/dao/dao.js:347]: Please fix indentation

File: src/components/dao/dao.js (hunk 5 above)
Comment (mohamedalichelbi) [reply src/components/dao/dao.js:344]: Just in case the `return;` is removed in the future

**Pull request merged**